phoenix-core/src/main/java/org/apache/phoenix/mapreduce/index/IndexUpgradeTool.java

```diff
@@ -0,0 +1,562 @@
+        indexingTool = new IndexTool();
+    }
+
+    @VisibleForTesting
+    public void prepareToolSetup() {
+        try {
+            if (logFile != null) {
+                FileHandler fh = new FileHandler(logFile);
+                fh.setFormatter(new SimpleFormatter());
+                LOGGER.addHandler(fh);
+            }
+
+            prop.put(Indexer.INDEX_BUILDER_CONF_KEY, PhoenixIndexBuilder.class.getName());
+            prop.put(NonTxIndexBuilder.CODEC_CLASS_NAME_KEY, PhoenixIndexCodec.class.getName());
+
+            if (inputTables == null) {
+                inputTables = new String(Files.readAllBytes(Paths.get(inputFile)));
+            }
+            if (inputTables == null) {
+                LOGGER.severe("Tables' list is not available; use -tb or -f option");
+            }
+            LOGGER.info("list of tables passed: " + inputTables);
+
+            if (operation.equalsIgnoreCase(UPGRADE_OP)) {
+                upgrade = true;
+            } else if (operation.equalsIgnoreCase(ROLLBACK_OP)) {
+                upgrade = false;
+            } else {
+                throw new IllegalStateException("Invalid option provided for "
+                        + OPERATION_OPTION.getOpt() + " expected values: {upgrade, rollback}");
+            }
+            if (dryRun) {
+                LOGGER.info("This is the beginning of the tool with dry run.");
+            }
+        } catch (IOException e) {
+            LOGGER.severe("Something went wrong "+e);
+            System.exit(-1);
+        }
+    }
+
+    @VisibleForTesting
+    public int executeTool() {
+        Configuration conf = HBaseConfiguration.addHbaseResources(getConf());
+
+        try (Connection conn = ConnectionUtil.getInputConnection(conf)) {
+
+            ConnectionQueryServices queryServices = conn.unwrap(PhoenixConnection.class)
+                    .getQueryServices();
+
+            boolean status = extractTablesAndIndexes(conn.unwrap(PhoenixConnection.class));
+
+            if (status) {
+                return executeTool(conn, queryServices, conf);
+            }
+        } catch (SQLException e) {
+            LOGGER.severe("Something went wrong in executing tool "+ e);
+        }
+        return -1;
+    }
+
+    private int executeTool(Connection conn, ConnectionQueryServices queryServices,
+            Configuration conf) {
+
+        LOGGER.info("Executing " + operation);
+
+        for (Map.Entry<String, HashSet<String>> entry : tablesAndIndexes.entrySet()) {
+            String dataTableFullName = entry.getKey();
+            HashSet<String> indexes = entry.getValue();
+
+            try (Admin admin = queryServices.getAdmin()) {
+
+                PTable dataTable = PhoenixRuntime.getTableNoCache(conn, dataTableFullName);
+                LOGGER.fine("Executing " + operation + " for " + dataTableFullName);
+
+                boolean mutable = !(dataTable.isImmutableRows());
+                if (!mutable) {
```

Comment:
@swaroopak - anything that would prevent an operator from running multiple calls to the tool at once? (e.g, if I have 100 data tables, can I run this 10 times in parallel from the command line to get all tables done faster?)

Reply:
Yes, it should be able to run them in parallel. In fact, using a parallel stream was an easier way to do it than dealing with multiple threads. Do you think we should implement multithreading here?

Reply:
If you can easily add multi-threading, go for it. (Remember though that for the 4.x and 4.14.x branches you can't use Java 8's pretty parallel stream APIs)